addons/mass_mailing/models/mailing.py

```diff
@@ -470,36 +470,13 @@ def _get_seen_list(self):
         self.ensure_one()
         target = self.env[self.mailing_model_real]
 
-        # avoid loading a large number of records in memory
-        # + use a basic heuristic for extracting emails
         query = """
-            SELECT lower(substring(t.%(mail_field)s, '([^ ,;<@]+@[^> ,;]+)'))
+            SELECT s.email
               FROM mailing_trace s
               JOIN %(target)s t ON (s.res_id = t.id)
-             WHERE substring(t.%(mail_field)s, '([^ ,;<@]+@[^> ,;]+)') IS NOT NULL
+             WHERE s.email IS NOT NULL
         """
 
-        # Apply same 'get email field' rule from mail_thread.message_get_default_recipients
-        if 'partner_id' in target._fields and target._fields['partner_id'].store:
```

Comment:
It seems the main failing use case if having a 'partner_id' field, as otherwise you have an email field and it is probably almost always using the right value. Could be worth a dedicated test in mass_mailing with its own test model. A lot of tests in mail use 'customer_id' as partner field (to try to find use case like this that are hardcoded and not really modular); but we lack a real model with 'partner_id' actually ^^ . In master we have a 'mailing.test.partner.unstored' model, could have a 'mailing.test.partner' model as well.